prow/create-build-cluster.sh

```diff
@@ -123,28 +139,125 @@ function createUploadSASecret() {
+    echo "gencred failed:" >&2
+    cat "$origdir/$outfile" >&2
+    return 1
+  )
+
+  # Store kubeconfig secret in the same project where build cluster is located
+  # and set up externalsecret in prow service cluster so that it's synced.
+  # First enable secretmanager API, no op if already enabled
+  gcloud services enable secretmanager.googleapis.com --project="${PROJECT}"
   cd "${origdir}"
-  echo
-  echo "Supply the file '${outfile}' to the current oncall for them to add to Prow's kubeconfig secret via:"
-  echo "  kubectl --context=<kubeconfig-context-for-prow-cluster> create secret generic kubeconfig-${clusterAlias} --from-file=kubeconfig=${outfile}"
-  echo "  Prow oncall should create PR updating build clusters consuming prow components to also include this new build cluster as part of KUBECONFIG env var."
-  echo "  An example of this PR is https://github.com/GoogleCloudPlatform/oss-test-infra/pull/653"
+  local gsm_secret_name="prow_build_cluster_kubeconfig_${clusterAlias}"
+  gcloud secrets create "${gsm_secret_name}" --data-file="$origdir/$outfile" --project="${PROJECT}"
+  # Grant prow service account access to secretmanager in build cluster
+  for role in roles/secretmanager.viewer roles/secretmanager.secretAccessor; do
+    gcloud beta secrets add-iam-policy-binding "${gsm_secret_name}" --member="serviceAccount:${PROW_SECRET_ACCESSOR_SA}" --role="${role}" --project="${PROJECT}"
+  done
+
+  prompt "Create CL for you" create_cl
+
   echo "ProwJobs that intend to use this cluster should specify 'cluster: ${clusterAlias}'" # TODO: color this
   echo
   echo "Press any key to acknowledge (this doesn't need to be completed to continue this script, but it needs to be done before Prow can schedule jobs to your cluster)..."
   pause
 }
+
+
+ensure_kustomize() {
+  if ! which "kustomize" > /dev/null 2>&1; then
+    GOBIN=$(pwd)/ GO111MODULE=on go get sigs.k8s.io/kustomize/kustomize/v3
+  fi
+}
+
+create_cl() {
+  cd "${ROOT_DIR}"
+  clone_uri="$(git config --get remote.origin.url)"
+  fork="$(echo "${clone_uri}" | gsed -e "s;https://github.com/;;" -e "s;git@github.com:;;" -e "s;.git;;")"
+  if [[ "${fork}" == "kubernetes/test-infra" ]]; then # This is not a fork
+    echo
+    read -r -n1 -p "Please provide the clone url for your fork: "
+    echo
+    clone_uri="$REPLY"
+    fork="$(echo "${clone_uri}" | gsed -e "s;https://github.com/;;" -e "s;git@github.com:;;" -e "s;.git;;")"
+  fi
+
+  local prow_deployment_dir="./config/prow/cluster"
+  cd "${tempdir}"
+  ensure_kustomize
+  git clone "${clone_uri}" --depth=1 forked-test-infra
+  cd forked-test-infra
+
+  git checkout -b add-build-cluster-secret
+
+  cat>>"${prow_deployment_dir}/kubernetes_external_secrets.yaml" <<EOF
+---
+apiVersion: kubernetes-client.io/v1
+kind: ExternalSecret
+metadata:
+  name: kubeconfig-build-${TEAM}
+  namespace: default
+spec:
+  backendType: gcpSecretsManager
+  projectId: ${PROJECT}
+  data:
+  - key: ${gsm_secret_name}
```

Comment:
This is a local variable of the caller so this won't be set, right?

Reply:
good catch! Didn't know how it worked in my local test, let me fix it

Reply:
https://github.com/kubernetes/test-infra/pull/21770